### PR TITLE
Move to file storage for log counting.

### PR DESCRIPTION
**Pull request by ryanwilson**
This prevents issues with writing to User Defaults on (or off) the main thread.

### CODE REVIEW

File: Firebase/Core/FIRLogger.m [hunk 1]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+  NSString *cacheDir = directories[0];
+  return [cacheDir stringByAppendingPathComponent:pathToAppend];
+}
+
+/**
+ * Read an integer from the specific file, returning kFIRLoggerCountFileNotFound if the file doesn't
+ * exist.
+ */
+NSInteger FIRReadIntegerFromFile(NSString *filePath) {
+  if (filePath == nil) {
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: !filePath.length

File: Firebase/Core/FIRLogger.m [hunk 2]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+  }
+
+  return [fileContents integerValue];
+}
+
+/**
+ * Read an integer from the specific file, returning YES if the write was successful.
+ */
+BOOL FIRWriteIntegerToFile(NSInteger value, NSString *filePath) {
+  if (filePath == nil) {
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: !filePath.length

File: Firebase/Core/FIRLogger.m [hunk 3]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+  if (error) {
+    FIRLogDebug(kFIRLoggerCore, @"I-COR000029",
+                @"Attempted to write the a log counter to file but failed: %@", error);
+  }
+
+  return success;
+}
+
+/**
+ * Return the number of errors logged since last being reset. Note: this synchronously reads a file
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: Returns the number ...

File: Firebase/Core/FIRLogger.m [hunk 4]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+
+  if (error) {
+    return 0;
+  }
+
+  return [fileContents integerValue];
+}
+
+/**
+ * Return the number of warnings logged since last being reset. Note: this synchronously reads a
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: Returns the number ...

File: Firebase/Core/FIRLogger.m [hunk 5]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+
+  if (error) {
+    return 0;
+  }
+
+  return [fileContents integerValue];
+}
+
+/**
+ * Reset the number of issues logged (warnings and errors).
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: Resets the number ...

File: Firebase/Core/FIRLogger.m (hunk 2 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m (hunk 3 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m (hunk 4 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m (hunk 1 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m (hunk 5 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m [hunk 6]
```diff
@@ -242,18 +352,26 @@ void FIRLogBasic(FIRLoggerLevel level,
...
-    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
-    NSInteger errorCount = [defaults integerForKey:kFIRLoggerErrorCountKey];
-    [defaults setInteger:errorCount + 1 forKey:kFIRLoggerErrorCountKey];
-  } else if (level == FIRLoggerLevelWarning) {
-    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
-    NSInteger warningCount = [defaults integerForKey:kFIRLoggerWarningCountKey];
-    [defaults setInteger:warningCount + 1 forKey:kFIRLoggerWarningCountKey];
-  }
+    // Keep count of how many errors and warnings are triggered.
+    if (level == FIRLoggerLevelError) {
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: Can we do merge the logic here?
NSString *path;
if (level == FIRLoggerLevelError) { 
  path = FIRLoggerPathInCachesByAppending(kFIRLoggerErrorCountFileName);
} else if (level == FIRLoggerLevelWarning) {
  path = FIRLoggerPathInCachesByAppending(kFIRLoggerWarningCountFileName);
}
NSInteger currentValue = FIRReadIntegerFromFile(path);
if (currentValue <= kFIRLoggerCountFileNotFound) {
    FIRWriteIntegerToFile(1, path);
} else {
     FIRWriteIntegerToFile(currentValue + 1, path);
}

Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Great idea, done!

File: Firebase/Core/FIRLogger.m [hunk 7]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
 
+#pragma mark - Number of errors and warnings
+
+NSString *FIRLoggerPathInCachesByAppending(NSString *pathToAppend) {
+  NSArray *directories =
+      NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
+  if (directories.count == 0) {
+    return nil;
+  }
+  NSString *cacheDir = directories[0];
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: To avoid crashing
NSString *cacheDir = directories.firstObject;

File: Firebase/Core/FIRLogger.m [hunk 8]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+  return [cacheDir stringByAppendingPathComponent:pathToAppend];
+}
+
+/**
+ * Read an integer from the specific file, returning kFIRLoggerCountFileNotFound if the file doesn't
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: Reads an integer ...

File: Firebase/Core/FIRLogger.m (hunk 7 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m (hunk 8 above)
Comment (ryanwilson) [reply Firebase/Core/FIRLogger.m]: Done.

File: Firebase/Core/FIRLogger.m [hunk 9]
```diff
@@ -213,6 +214,115 @@ BOOL getFIRLoggerDebugMode() {
...
+  }
+
+  return success;
+}
+
+/**
+ * Returns the number of errors logged since last being reset. Note: this synchronously reads a file
+ * on the same queue that logging occurs.
+ */
+NSInteger FIRNumberOfErrorsLogged(void) {
```
Comment (baolocdo) [inline Firebase/Core/FIRLogger.m]: Is this only for testing? I can't find it's used in other places. How often are we reading this synchronously? I'm afraid of blocking the caller, especially on the main queue.

**Comment by ryanwilson** [comment]
Removing from this release in #1198 - will be revisiting this for a later release.

**Comment by ryanwilson** [comment]
Closing due to inactivity.

**Pull request closed without merging**